components/Buttons/src/MDCFloatingButton.m

```diff
@@ -72,67 +105,406 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  }
+
+  if (self.minimumSize.height > 0) {
+    contentSize.height = MAX(self.minimumSize.height, contentSize.height);
+  }
+  if (self.maximumSize.height > 0) {
+    contentSize.height = MIN(self.maximumSize.height, contentSize.height);
+  }
+  if (self.minimumSize.width > 0) {
+    contentSize.width = MAX(self.minimumSize.width, contentSize.width);
+  }
+  if (self.maximumSize.width > 0) {
+    contentSize.width = MIN(self.maximumSize.width, contentSize.width);
   }
+  return contentSize;
 }
 
 - (void)layoutSubviews {
   // We have to set cornerRadius before laying out subviews so that the boundingPath is correct.
   self.layer.cornerRadius = CGRectGetHeight(self.bounds) / 2;
   [super layoutSubviews];
+
+   if (self.mode == MDCFloatingButtonModeNormal) {
+    return;
+  }
+
+  // Position the imageView and titleView
+  //
+  // +------------------------------------+
+  // |    |  |  |  CEI TOP            |   |
+  // |CEI +--+  |       +-----+       |CEI|
+  // | LT ||SP|-- A --|Title|-- A --|RGT|
+  // |    +--+  |       +-----+       |   |
+  // |    |  |  |  CEI BOT            |   |
+  // +------------------------------------+
+  //
+  // (A) The same spacing on either side of the label.
+  // (SP) The spacing between the image and title
+  // (CEI) Content Edge Insets
+  //
+  // The diagram above assumes an LTR user interface orientation
+  // and a .leadingIcon imageLocation for this button.
+
+  BOOL isLeadingIcon = self.imageLocation == MDCFloatingButtonImageLocationLeading;
+  UIEdgeInsets adjustedLayoutInsets =
+      isLeadingIcon ? internalLayoutSpacingInsets
+                    : MDFInsetsFlippedHorizontally(internalLayoutSpacingInsets);
+
+  const CGRect insetBounds = UIEdgeInsetsInsetRect(UIEdgeInsetsInsetRect(self.bounds,
+                                                                         adjustedLayoutInsets),
+                                                   self.contentEdgeInsets);
+
+  const CGFloat imageViewWidth = CGRectGetWidth(self.imageView.bounds);
+  const CGFloat boundsCenterY = CGRectGetMidY(insetBounds);
+  CGFloat titleWidthAvailable = CGRectGetWidth(insetBounds);
+  titleWidthAvailable -= imageViewWidth;
+  titleWidthAvailable -= self.imageTitleSpacing;
+
+  const CGFloat availableHeight = CGRectGetHeight(insetBounds);
+  CGSize titleIntrinsicSize
+      = [self.titleLabel sizeThatFits:CGSizeMake(titleWidthAvailable, availableHeight)];
+
+  const CGSize titleSize = CGSizeMake(MAX(0, MIN(titleIntrinsicSize.width, titleWidthAvailable)),
+                                      MAX(0, MIN(titleIntrinsicSize.height, availableHeight)));
+
+  CGPoint titleCenter;
+  CGPoint imageCenter;
+  BOOL isLTR =
+      self.mdf_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionLeftToRight;
+
+  // If we are LTR with a leading image, the image goes on the left.
+  // If we are RTL with a trailing image, the image goes on the left.
+  if ((isLTR && isLeadingIcon) || (!isLTR && !isLeadingIcon)) {
```

Comment:
this is an XOR? consider ^. https://stackoverflow.com/questions/11431844/xor-objective-c

Reply:
This is based on earlier feedback from @ianegordon. Using XOR is other operations can be unintuitive to maintainers so I think it's better to leave it as explicit.

Reply:
sgtm